### PR TITLE
KAFKA-9767: Add logging to basic auth rest extension

### PR DESCRIPTION
**Pull request by C0urante**
[Jira](https://issues.apache.org/jira/browse/KAFKA-9767)

The changes here are strictly targeted towards improving logging for the basic auth rest extension. However, in several sections, some refactoring is also performed in order to reduce nesting especially with the `else` halves of `if`/`else` statements.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by C0urante** [comment]
@mageshn, @rhauch would either of you mind taking a look when you have a moment?

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java
Comment (mageshn) [inline connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java:106]: This log is likely a duplicate of the one that we have in the `initialize`. Should we log this for every login?

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java
Comment (mageshn) [inline connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java:114]: do we really need that we matched it? Can't we do all all of this is a single log statement? We can include size of credentials map and authenticated boolean. This will help keep the old structure.

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java
Comment (mageshn) [inline connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java]: Any reason why we want to iterate all the call backs and then throw an exception?

**Review by mageshn — COMMENTED** [review]
@C0urante Thanks a lot for the PR. The logging improvements will generally be useful and overall LGTM. I had a couple of minor comments and a question before I give my +1 on this :)

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java
Comment (C0urante) [reply connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java]: Yes, this adds additional logging to let us see a complete list of the invalid callbacks.

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java
Comment (C0urante) [reply connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java:106]: I think it's alright if it's at `TRACE` level. My thinking was that since some Connect clusters are left running for a long time, the original log line in `initialize` may be lost by the time someone wants to investigate the behavior of the extension.

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java
Comment (C0urante) [reply connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java:114]: I think at `TRACE` level it makes sense to be more generous with logging; if the structure is unreadable because of the changes here I guess we can modify things but it still seems legible to me and might save someone a headache down the road when trying to debug issues with the extension.

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java
Comment (mageshn) [inline connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java]: Should this be `warn`. Its likely an incorrect set up that can potentially cause this and might be useful to know it by default.

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java
Comment (mageshn) [reply connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java:106]: Typically, most users push logs to something like elastic/splunk and should be able to lookup the logs from the archive. I'm not too concerned about this since properties file based credentials store shouldn't typically be used in a production environment. So,I'm fine leaving the log statement in here. But let's see what the committers have to say fro this :).

**Review by mageshn — COMMENTED** [review]
Thanks @C0urante for the response. I had one last question about the log level in one scenario.

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java
Comment (C0urante) [reply connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java:106]: I think it's a pretty big assumption that users will have infinite (or at least very forgiving) log retention, though, and banking on that leaves ones who have more conservative policies out on their own.

Additionally, the log message in `initialize` would be sufficient for anyone who knew to look for it, but for anyone else, a log message that happens whenever a REST request that gets authenticated would be easier to spot.

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java
Comment (C0urante) [reply connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java]: Sure, will do.

**Review by mageshn — APPROVED** [review]
Thanks a lot @C0urante . LGTM.

**Comment by C0urante** [comment]
Thanks @mageshn! @rhauch, @kkonstantine, would either of you mind taking a look when you have a chance?

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java
Comment (rhauch) [inline connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java]: Would it help to actually list the method that was used, in case somebody thought they were using basic?
```suggestion
                log.trace("Request credentials used {} authentication, but only {} supported; ignoring", BASIC, method);
```

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java
Comment (rhauch) [inline connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java]: The comment above says:
> and it is likely that the worker will need to be reconfigured and restarted

Should the log message say this? Something more like:
> Unsupported callback '{}' of type {}; request authentication will fail. This indicates the Connect worker was configured incorrectly, and will need to be corrected and restarted.

The more specific we could be about what might have been misconfigured, the better.

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java
Comment (rhauch) [inline connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java]: Here we're only reporting the last of potentially multiple unsupported callbacks, right? Should we be generating a list of unsupported callbacks here if we're waiting to throw the exception?

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java
Comment (rhauch) [inline connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java]: Doesn't this catch handle all exceptions, including authentication failures? Do we really want to log all of those auth failures at WARN?

**Review by rhauch — COMMENTED** [review]
Thanks, @C0urante. This seems really useful. I have a few nits and suggestions below.

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java
Comment (C0urante) [reply connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java]: I wanted to preserve the content of the original exception message here. If it's fine to change it, then yes, it'd definitely be an improvement to include the whole list here.

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java
Comment (C0urante) [reply connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/PropertyFileLoginModule.java]: I don't think this is going to get logged for auth failures (assuming you mean things like incorrect username/password, bad auth header, etc.).

You can see the code for `callbackHandler::handle` in the `JaasBasicAuthFilter.BasicAuthCallBackHandler` class; the only expected exception would be an `UnsupportedCallbackException`, which probably does warrant logging at `WARN` level.

I'd like to refactor this slightly to take into account your feedback about the logging/exception logic in `JaasBasicAuthFilter.BasicAuthCallBackHandler::handle`; I think we can keep this log message here, and use it in lieu of the `WARN`-level messages in the other file.

**Comment by C0urante** [comment]
Thanks for the review, @rhauch. After reading your comments and trying to put them into practice, I noticed a few additional areas where it seemed improvements could be made; let me know how this looks.
I've also tested locally and confirmed that no `WARN`-level messages are emitted when malformed or invalid credentials are provided.

**Comment by rhauch** [comment]
ok to test

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java
Comment (rhauch) [inline connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java:122]: Isn't it true that most of the time this method will get only supported `Callback` implementations? In those cases, we'll never need the `unsupportedCallbacks`. 

I originally thought it might be worth making this more efficient, but I don't think it's worth it since the default constructor for `ArrayList` is pretty efficient in recent JVMs (with lazy allocation of the underlying array when the first element is added to the list).

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java
Comment (rhauch) [inline connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java:143]: Do we have a unit test that exercises this code path? If not, what do you think about adding new unit tests for `BasicAuthCallBackHandler`?

**Review by rhauch — COMMENTED** [review]
Thanks, @C0urante. We actually don't have unit tests for `BasicAuthCallBackHandler`, and it'd be helpful to add some tests since we are refactoring the logic while trying to avoid changing behavior (other than adding logging).

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java
Comment (C0urante) [reply connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java:122]: 👍 happy to optimize if you'd like, but it sounds like we can leave it as-is for now at least.

File: connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java
Comment (C0urante) [reply connect/basic-auth-extension/src/main/java/org/apache/kafka/connect/rest/basic/auth/extension/JaasBasicAuthFilter.java:143]: Ack, added a new test for an unsupported callback. We have tests already for the green path, so it didn't seem necessary to add more than that.

**Comment by C0urante** [comment]
Thanks @rhauch; I've added the requested unit test. This is ready for another round when you have time

**Review by rhauch — APPROVED** [review]
LGTM, pending a green build. Thanks, @C0urante!

**Comment by rhauch** [comment]
retest this please

**Comment by rhauch** [comment]
ok to test

**Comment by rhauch** [comment]
ok to test

**Comment by rhauch** [comment]
One green build, and two builds with only a flaky Streams test failure.

**Pull request merged**
